Add directory upload support to st.chat_input

sfc-gh-smohile opened the pull request:
## Describe your changes

This PR adds directory upload support to st.chat_input, allowing users to select and upload entire directories in chat interfaces.

### Key Changes
- **Python API**: Added accept_files parameter that accepts 'directory' option
- **Frontend**: Added webkitdirectory support with client-side file type filtering
- **UX**: Prevents upload attempts for invalid file types (filters before upload)
- **Structure**: Preserves directory paths using webkitRelativePath

### Usage
```python
# Upload entire directory in chat
chat_input = st.chat_input('Type a message...', accept_files='directory')

# Upload directory with file type filter
chat_input = st.chat_input('Type a message...', 
                          accept_files='directory',
                          type=['png', 'jpg'])
```

## Related PRs
- Shares common utilities with #11905 (file_uploader directory support)

## Testing Plan
**Unit Tests**
- ✅ File filtering logic
- ✅ UI state changes 
- ✅ Protobuf field assignment
- ✅ Chat input with directory upload

**E2E Tests**
- ✅ Complete directory upload flow
- ✅ Error handling for invalid file types
- ✅ Directory structure preservation
- ✅ Browser compatibility testing

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-smohile committed:
Add directory upload support to st.chat_input

- Added accept_files parameter with 'directory' option
- Frontend support for webkitdirectory with file type filtering
- Preserves directory structure using webkitRelativePath
- Common utilities shared with file_uploader implementation

sfc-gh-bnisco reviewed (commented):
Just doing a quick run-down of this PR, I think there will be a lot of duplicated comments between this one and https://github.com/streamlit/streamlit/pull/11905 given the common code.

I think it makes sense to get 11905 in a good state and checked-in first. After, you can handle the merge conflicts in this branch, and then we'll do a proper review of this one.

sfc-gh-smohile commented:
> Just doing a quick run-down of this PR, I think there will be a lot of duplicated comments between this one and #11905 given the common code.
> 
> I think it makes sense to get 11905 in a good state and checked-in first. After, you can handle the merge conflicts in this branch, and then we'll do a proper review of this one.

That sounds good to me

sfc-gh-smohile committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Apply PR #11905 feedback patterns to chat input directory upload

- Improve file type validation to check extension boundaries properly
- Remove orphaned comment
- Align chat input code with approved file_uploader patterns

sfc-gh-smohile committed:
Same changes

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-nbellante commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-directory_filtered[chromium].png:
this snapshot seems truncated?

sfc-gh-nbellante commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-submit_hover[firefox].png:
where'd the 5/200 go?

sfc-gh-nbellante commented on e2e_playwright/st_chat_input.py:
```diff
@@ -58,15 +58,28 @@ def on_submit():
 
-v6 = st.container().chat_input("Chat input 7 (width=300px)", width=300)
-v7 = st.container().chat_input("Chat input 8 (width='stretch')", width="stretch")
+# Add directory upload test scenario
```
not related to this but why was v6 going to v7 and chatInput 6 not existing? 🤔

sfc-gh-bnisco replied:
**suggestion:** Remove this unnecessary LLM comment

```suggestion
```

sfc-gh-smohile committed:
Delete snapshots

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/ChatFileUploadButton.tsx:
```diff
@@ -47,34 +47,48 @@ const ChatFileUploadButton = ({
+              ? "files"
+              : acceptFile === AcceptFileValue.Directory
+                ? "a directory"
+                : "a file"
```
**suggestion:** Move this out into a construct that isn't a nested ternary for readability. Perhaps a standalone function or `useMemo`

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/ChatFileUploadDropzone.tsx:
```diff
@@ -37,17 +37,31 @@ const ChatFileUploadDropzone = ({
+  if (acceptFile === AcceptFileValue.Directory) {
+    inputProps.webkitdirectory = ""
+    inputProps.multiple = true
+  }
```
**suggestion:** Seeing how the logic changes in this file are 95% the same as in `ChatFileUploadButton`, I suggest we move this logic into its own shared place. Could be a custom hook or just a standalone function that encapsulated creating the input props and the nested ternary logic for the copy.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/createDropHandler.ts:
```diff
@@ -17,35 +17,106 @@
+    )
+  })
+}
+
```
+1 to this suggestion, this type of logic definitely needs to be shared to prevent subtle divergent behavior changes over time.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+
+    // Verify aria labels and accessibility
+    const chatInput = screen.getByTestId("stChatInputTextArea")
+    expect(chatInput).toBeInTheDocument()
```
**suggestion:** This is a redundant assertion, please utilize something more meaningful such as `toBeVisible`.

sfc-gh-smohile committed:
UPdated snapshots

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+  })
+
+  it("filters directory files by allowed types", () => {
+    const consoleSpy = vi.spyOn(console, "log")
```
**suggestion:** Much like the other PR, please remove these unused console spys

sfc-gh-bnisco replied:
Bumping this again, I think this was accidentally reverted

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+    expect(submitButton).toBeEnabled()
+  })
+
+  it("displays directory upload instructions correctly", () => {
```
**issue:** This test doesn't seem to assert that upload instructions are correct, please update the test to reflect the expected behavior.

sfc-gh-bnisco replied:
Bumping this one as well

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+    expect(fileInput).toHaveAttribute("multiple")
+  })
+
+  it("removes directory files when deleted individually", () => {
```
**issue:** This test doesn't seem to assert anything about removing directory files, please update the test to reflect the expected behavior.

sfc-gh-bnisco replied:
Bumping this one as well

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+    await user.type(chatInput, "Here are the project files")
+
+    // Mock directory files with relative paths
+    const directoryFiles = [
+      Object.assign(
+        new File(["main content"], "main.py", { type: "text/plain" }),
+        {
+          webkitRelativePath: "project/main.py",
+        }
+      ),
+      Object.assign(
+        new File(["test content"], "test.py", { type: "text/plain" }),
+        {
+          webkitRelativePath: "project/tests/test.py",
+        }
+      ),
+    ]
```
**suggestion:** Much like the last PR, we moved this logic into a shared helper function. This PR should re-use that same logic by placing it into a shared location and using it in all test files that need to create Files

sfc-gh-bnisco replied:
Bumping this one

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +428,278 @@ describe("ChatInput widget", () => {
+
+    render(<ChatInput {...props} />)
+
+    // Since we can't easily simulate file upload in unit tests,
+    // let's verify the component is set up correctly for directory uploads
```
**issue:** There is an inconsistency in what this comment says and what tests later in this file are doing - it looks like we actually simulating file upload and asserting against `chatInputValue.fileUploaderState`. These tests should be updated to also actually upload something.

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-long_placeholder[firefox].png:
**issue:** This looks to be targeting the wrong element now

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-app_bottom[chromium].png:
**issue:** The chat input numbers changing for existing tests is an unnecessary change that causes a lot of extra diffs in the snapshots. Please maintain the existing indices in the test file, then undo these snapshot changes to keep the git diff more focused.

sfc-gh-smohile committed:
Snapshot changes, removed LLM comments, moved nested ternary, Created shared utilities in fileUploadUtils.ts

sfc-gh-smohile committed:
Snapshots

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/fileUploadUtils.ts:
```diff
@@ -0,0 +1,74 @@
+    case AcceptFileValue.Directory:
+      return "a directory"
+    default:
+      return "a file"
```
**suggestion:** To increase the usefulness of the compiler, utilize `assertNever(acceptFile)` here. That way, if the type of `AcceptFileValue` ever includes other values, the TypeScript compiler will let us know we need to update this switch statement.

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
fix lint error

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Updated test

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Resolve merge conflict in ChatInput upload handler: keep AbortController flow, preserve directory webkitRelativePath for filenames

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Removed orphanned snapshots

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Updated snapshots

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on e2e_playwright/st_chat_input.py:
```diff
@@ -61,12 +61,17 @@ def on_submit():
 v7 = st.container().chat_input("Chat input 8 (width='stretch')", width="stretch")
 
+# Add directory upload test scenario
+v8 = st.container().chat_input(
+    "Chat input 9 (directory)", accept_file="directory", file_type=["txt", "py", "md"]
+)
+st.write("Chat input 9 (directory) - value:", v8)
```
**suggestion:** This should probably be called `v9` to keep it aligned with the text. It seems like v10 is aligned with `Chat input 10`, although I understand others in this file are still misaligned. It would be good for new code here to showcase clear behavior.

sfc-gh-bnisco commented on e2e_playwright/st_file_uploader_test.py:
```diff
@@ -365,10 +365,14 @@ def test_directory_upload_with_file_type_filtering(
     temp_dir = create_temp_directory_with_files(directory_data)
 
-    with app.expect_file_chooser() as fc_info:
+    with app.expect_file_chooser(timeout=5000) as fc_info:
```
**question:** Why is this `timeout` necessary? What fails without it?

sfc-gh-bnisco commented on e2e_playwright/st_file_uploader_test.py:
```diff
@@ -365,10 +365,14 @@ def test_directory_upload_with_file_type_filtering(
 
     file_chooser = fc_info.value
+
+    # Add a small delay before setting files to ensure file chooser is ready
+    app.wait_for_timeout(100)
```
**question:** Is there a better way to ensure the file chooser is ready? Waiting for a specific period of time should be avoided if possible.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/createDropHandler.ts:
```diff
@@ -17,35 +17,101 @@
+        file,
+        errors: [
+          {
+            code: "file-invalid-type",
```
+1 to this

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/createDropHandler.ts:
```diff
@@ -17,35 +17,85 @@
 import { FileUploadClient } from "~lib/FileUploadClient"
 import { getRejectedFileInfo } from "~lib/util/FileHelper"
 
+import { isFileTypeAllowed as checkFileType } from "./fileUploadUtils"
```
**suggestion:** Since this is a named export, utilize the named export unless there is a good reason to alias it.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/createFileUploadHandler.ts:
```diff
@@ -42,9 +42,15 @@ export const createUploadFileHandler =
+    // For directory uploads, prefer the webkitRelativePath so we preserve
+    // the original directory structure in the displayed file name.
+    const fileName =
+      (file as File & { webkitRelativePath?: string }).webkitRelativePath ||
+      file.name
```
**suggestion:** This type casting shouldn't be necessary. `const fileName = file.webkitRelativePath || file.name` should suffice.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/fileUploadUtils.ts:
```diff
@@ -0,0 +1,79 @@
+    case AcceptFileValue.Directory:
+      return "a directory"
+    default:
+      return assertNever(acceptFile)
```
+1 to this

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-shift_enter_new_line[chromium].png:
**issue:** It seems like this is missing the character count display now. That is unexpected.

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Addressed copilot feedback

sfc-gh-smohile committed:
Addressed Bob's feedback

sfc-gh-smohile committed:
Frontend tests

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Removed timeouts

sfc-gh-smohile committed:
Moved mouse

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-shift_enter_new_line[chromium].png:
**issue:** This snapshot is missing the current character count in the bottom right

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-shift_enter_new_line[firefox].png:
**issue:** This snapshot is missing the current character count in the bottom right

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-shrinks[firefox].png:
**issue:** This snapshot is missing the current character count in the bottom right

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-shrinks[chromium].png:
**issue:** This snapshot is missing the current character count in the bottom right

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-focused[chromium].png:
**issue:** This test is targeting the incorrect chat input now since some of them have been reordered. Please ensure this test continues to target the same input as before this change.

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-directory_uploaded[chromium].png:
**issue:** Based off of what we learned in https://github.com/streamlit/streamlit/pull/12188 - we shouldn't have snapshot tests for uploaded files since there is no guarantee in order. Remove these and update the tests to just assert that the expected files are there in text.

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Updated a lot of snapshots, deleted directory snapshots, will updated failed snapshots soon

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Update snapshots again

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Reset st_chat_input tests and snapshots to develop branch

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Revert snapshots to develop, revert removed unit tests

sfc-gh-smohile committed:
Update frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

sfc-gh-smohile committed:
Updated a bunch of snapshots, updated tests

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Update st_chat_input_test.py

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/fileUploadUtils.ts:
```diff
@@ -0,0 +1,80 @@
+  return acceptedExtensions.some(ext => {
+    const extLower = ext.toLowerCase()
+    // Ensure we're matching at an extension boundary (preceded by a period or starting with one)
+    return (
+      fileName.endsWith(extLower) &&
+      (extLower.startsWith(".") || fileName.endsWith(`.${extLower}`))
+    )
```
+1 to this and the copilot comments here, looks like we have some logic bugs we definitely need to solve here!

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Updated order for snapshots

sfc-gh-smohile committed:
fix: Fix test indices and code review issues for chat input directory upload

- Update st_chat_input_test.py to correct DOM ordering (bottom chat input at index 9)
- Fix boolean logic in chat.py using set membership check
- Simplify file extension matching logic in fileUploadUtils.ts
- Extract shared file validation logic to reduce duplication
- Fix test naming to focus on user behavior

sfc-gh-smohile committed:
Updated snapshots

sfc-gh-smohile committed:
Fixed broken tests and implemented feedback

sfc-gh-smohile committed:
Modified tooltip tests

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/fileUploadUtils.ts:
```diff
@@ -0,0 +1,80 @@
+    return (
+      fileName.endsWith(extLower) &&
+      (extLower.startsWith(".") || fileName.endsWith(`.${extLower}`))
+    )
```
+1 to these bot comments, let's address this suggestion in this PR and write unit tests to assert this behavior.

sfc-gh-bnisco reviewed (commented):
Snapshots look good, a few more things in line.

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Addressed reverted feedback

sfc-gh-smohile committed:
Merge branch 'develop' into feature/upload_dir_chat_input

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -428,4 +432,311 @@ describe("ChatInput widget", () => {
+      expect(fileNames).toHaveLength(2)
+      expect(fileNames[0]).toHaveTextContent("folder/file1.txt")
+      expect(fileNames[1]).toHaveTextContent("folder/file2.txt")
+    })
```
**question:** Is there any chance this will flake due to upload order? I know this is a unit test, but I'm not sure if there is a guarantee on upload order in this semi-mocked environment

sfc-gh-smohile replied:
Good eye! Let me change that.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/fileUploadUtils.test.ts:
```diff
@@ -0,0 +1,187 @@
+describe("fileUploadUtils", () => {
+  describe("isFileTypeAllowed", () => {
+    it("allows all files when no extensions are specified", () => {
+      const file = new File(["content"], "test.txt", { type: "text/plain" })
```
**suggestion:** All of these files should be using the newly created `createFileWithPath` test helper

sfc-gh-smohile committed:
Fixed failing frontend test

sfc-gh-smohile committed:
Merge branch 'feature/upload_dir_chat_input' of https://github.com/streamlit/streamlit into feature/upload_dir_chat_input

sfc-gh-smohile committed:
Prevented flakiness and refractored code

sfc-gh-bnisco reviewed (approved):
Thank you for the iterations on this! LGTM

sfc-gh-smohile commented:
> Thank you for the iterations on this! LGTM

Woo hoo!
